driver-async/src/main/com/mongodb/async/client/internal/KeyManagementService.java

```diff
@@ -0,0 +1,117 @@
+import com.mongodb.connection.TlsChannelStreamFactoryFactory;
+import com.mongodb.crypt.capi.MongoKeyDecryptor;
+import org.bson.ByteBuf;
+import org.bson.ByteBufNIO;
+
+import javax.net.ssl.SSLContext;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+class KeyManagementService {
+    private final int port;
+    private final StreamFactory streamFactory;
+
+    KeyManagementService(final SSLContext sslContext, final int port, final int timeoutMillis) {
+        this.port = port;
+        this.streamFactory = new TlsChannelStreamFactoryFactory().create(SocketSettings.builder()
+                        .connectTimeout(timeoutMillis, TimeUnit.MILLISECONDS)
+                        .readTimeout(timeoutMillis, TimeUnit.MILLISECONDS)
+                        .build(),
+                SslSettings.builder().enabled(true).context(sslContext).build());
+    }
+
+    void decryptKey(final MongoKeyDecryptor keyDecryptor, final SingleResultCallback<Void> callback) {
+        streamOpen(keyDecryptor, callback);
+    }
+
+    private void streamOpen(final MongoKeyDecryptor keyDecryptor, final SingleResultCallback<Void> callback) {
+        final Stream stream = streamFactory.create(new ServerAddress(keyDecryptor.getHostName(), port));
+        stream.openAsync(new AsyncCompletionHandler<Void>() {
+            @Override
+            public void completed(final Void aVoid) {
+                streamWrite(stream, keyDecryptor, callback);
+            }
+
+            @Override
+            public void failed(final Throwable t) {
+                stream.close();
+                callback.onResult(null, new MongoSocketOpenException("Exception opening connection to Key Management Service",
+                        getServerAddress(keyDecryptor), t));
+            }
+        });
+    }
+
+    private void streamWrite(final Stream stream, final MongoKeyDecryptor keyDecryptor, final SingleResultCallback<Void> callback) {
+        List<ByteBuf> byteBufs = Collections.<ByteBuf>singletonList(new ByteBufNIO(keyDecryptor.getMessage()));
+        stream.writeAsync(byteBufs, new AsyncCompletionHandler<Void>() {
+            @Override
+            public void completed(final Void aVoid) {
+                streamRead(stream, keyDecryptor, callback);
+            }
+
+            @Override
+            public void failed(final Throwable t) {
+                stream.close();
+                callback.onResult(null, new MongoSocketWriteException("Exception sending message to Key Management Service",
+                        getServerAddress(keyDecryptor), t));
+            }
+        });
+    }
+
+    private void streamRead(final Stream stream, final MongoKeyDecryptor keyDecryptor, final SingleResultCallback<Void> callback) {
+        int bytesNeeded = keyDecryptor.bytesNeeded();
+        if (bytesNeeded > 0) {
+            bytesNeeded = bytesNeeded > 230 ? 230 : bytesNeeded; // TODO - why report 1024 needed and only supply 230?
```

Comment:
So this caused quite a debugging headache.  Even though keyDecryptor reports it needs 1024 bytes on the first call, it only supplies 230.  That is not an issue with the sync stream but with async it causes a `MongoSocketReadException: Prematurely reached end of stream` exception.

I debugged the sizes coming through on sync and found that 230 was the first return and fixed the max size. Once set it all started working :/

I'm not sure on the correct action here going forward or if this hack will suffice.

Reply:
Done - added a wrapper to access the channel directly

Reply:
I find the solution just a little unintuitive.  The contract of readAsync says: "This method should ... invoke the given callback when the number of requested bytes have been read."  So I don't love that the implementation of it in the new class violates that contract.  

Since the caller is taking a dependency on the concrete KeyManagementAsynchronousChannelStream type, would you consider instead adding a new method to that type with the behavior that we require here? What to name that method though... perhaps readAtMost?  readAtMostBytes?

Reply:
Updated so we just use the channel directly in the KeyManagementService